### PR TITLE
automated release

### PR DESCRIPTION
**Pull request by kubawach**
Automates lambda release - no need for manual steps anymore.

The only thing that has not been tested is actual git push. Perhaps it will be necessary to use https://github.com/marketplace/actions/push-to-master or similar.

### CODE REVIEW

**Comment by pellared** [comment]
> The only thing that has not been tested is actual git push. Perhaps it will be necessary to use https://github.com/marketplace/actions/push-to-master or similar.

You can test it on a fork 😉

**Comment by kubawach** [comment]
> > The only thing that has not been tested is actual git push. Perhaps it will be necessary to use https://github.com/marketplace/actions/push-to-master or similar.
> 
> You can test it on a fork 😉

Sometimes the easiest solutions are the hardest to spot :D Thanks for the suggestion.

File: wrapper/pom.xml [hunk 1]
```diff
@@ -47,6 +47,122 @@
...
+                <version>3.0.0-M4</version>
+                <configuration>
+                    <pushChanges>false</pushChanges>
+                </configuration>
+            </plugin>
+            <plugin>
+                <artifactId>exec-maven-plugin</artifactId>
+                <groupId>org.codehaus.mojo</groupId>
+                <configuration>
+                    <executable>./.github/update-version.sh</executable>
```
Comment (mateuszrzeszutek) [inline wrapper/pom.xml:79]: Does this script exist somewhere?

File: .github/workflows/release.yaml [hunk 2]
```diff
@@ -20,13 +38,30 @@ jobs:
...
-      - name: Build wrapper with Maven
-        id: build_wrapper
-        run: mvn -B clean package --file ./wrapper/pom.xml
-
       - name: Get release version
         id: get_version
-        run: echo ::set-output name=VERSION::${GITHUB_REF:11}
+        run: echo ::set-output name=VERSION::${GITHUB_REF:12}
+
+      - name: Create new devlopment version
```
Comment (mateuszrzeszutek) [inline .github/workflows/release.yaml:45]: Typo: 
```suggestion
      - name: Create new development version
```

File: pre-release.sh [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+next_release=$1
+
+echo "Current release: $current_release"
+echo "Current development: $current_dev"
+echo "Next release: $next_release"
+
+mvn versions:set -DnewVersion="$next_release" -DgenerateBackupPoms=false -f wrapper/pom.xml
+sed -i "s/$current_release/$next_release/" ./README.md
+sed -i "s/$current_release/$next_release/" ./outbound-context-propagation.md
+sed -i -r "s/(wrapper:\s')$current_dev/\1$next_release/" ./examples/build.gradle
```
Comment (mateuszrzeszutek) [inline pre-release.sh:24]: Nice

File: .github/workflows/release.yaml [hunk 4]
```diff
@@ -41,8 +76,8 @@ jobs:
           name: Release ${{ steps.get_version.outputs.VERSION }}
           files: otel-java-lambda-wrapper.jar
```
Comment (mateuszrzeszutek) [inline .github/workflows/release.yaml:77]: Consider adding `draft: true` to have a chance to edit release notes before publishing.

File: post-release.sh [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+
+if [[ ! "$#" -eq 2 ]]
+then
+  echo "usage: $0 NEXT_RELEASE_VERSION NEXT_DEVELOPMENT_VERSION"
+  exit 1
+fi
+
+current_release=`cat release.properties | grep "current.release=" | cut -c17-`
+next_dev=$2
+next_release=$1
```
Comment (mateuszrzeszutek) [inline post-release.sh:17]: At this point in time `current_release` is actually the previous release, and `next_release` is the current one. How about renaming the env vars?

File: RELEASING.md [hunk 6]
```diff
@@ -1,22 +1,29 @@
...
+  * creates GitHub release
+  * performs M2 release
+  * updates POM / examples to next development version
+  * updates release properties
+  * commits
+  * pushed all changes
 
 After the release is completed successfully, the following should happen:
-* Github tag / release created
+* Github release tag (`r<VERSION>`) created
```
Comment (mateuszrzeszutek) [inline RELEASING.md:24]:
```suggestion
* Github release tag (`v<VERSION>`) created
```

File: .github/workflows/release.yaml [hunk 7]
```diff
@@ -20,13 +38,30 @@ jobs:
...
+
+      - name: Configure git
+        id: configure_git
+        run: |
+          git config --global user.name 'Lambda wrapper CI'
+          git config --global user.email 'lambda-wrapper-ci@users.noreply.github.com'
+
+      - name: Prepare release
+        id: prepare_release
+        run: ./pre-release.sh ${{ steps.get_version.outputs.VERSION }}
```
Comment (mateuszrzeszutek) [inline .github/workflows/release.yaml:60]: WDYT about hiding those `(pre|post)-release.sh` files in the `.github` directory? They're not supposed to be used by any human being directly, they're tightly related to the release workflow.

File: RELEASING.md [hunk 8]
```diff
@@ -1,22 +1,29 @@
...
-* Tag that version (eg `v0.0.5` - mind `v` prefix!) with the [annotated tag](https://git-scm.com/book/en/v2/Git-Basics-Tagging) and `git push origin <tagname>` 
-* tag will run the release workflow that needs to be *approved* by a maintainer 
-* Update wrapper POM version to next `-SNAPSHOT` version
-* Set wrapper dependency for examples (`examples/build.gradle`) to next `-SNAPSHOT` version
-* Commit / push / merge PR
+* Tag the release candidate version (eg `rc0.0.5`) with the [annotated tag](https://git-scm.com/book/en/v2/Git-Basics-Tagging) and `git push origin <tagname>` 
+* Release workflow will be run, it needs to be *approved* by a maintainer 
+* Release will perform following actions:
+  * updates POM / examples / docs version to requested release
+  * commits
```
Comment (mateuszrzeszutek) [inline RELEASING.md:13]: I wasn't sure what "commits" referred to at the first look, how about slightly rewording this list:

```suggestion
  * update POM / examples / docs version to requested release
  * commit changes
```

File: wrapper/pom.xml (hunk 1 above)
Comment (kubawach) [reply wrapper/pom.xml:79]: Leftovers from previous attempt with maven-release plugin. I shall remove whole config.

File: pre-release.sh (hunk 3 above)
Comment (kubawach) [reply pre-release.sh:24]: Probably something that next maintainer of this file will hate ;)

File: .github/workflows/release.yaml (hunk 4 above)
Comment (kubawach) [reply .github/workflows/release.yaml:77]: Good idea!

File: post-release.sh (hunk 5 above)
Comment (kubawach) [reply post-release.sh:17]: Done. Also commented line where current becomes previous.

File: RELEASING.md (hunk 6 above)
Comment (kubawach) [reply RELEASING.md:24]: Nice catch.

File: RELEASING.md (hunk 8 above)
Comment (kubawach) [reply RELEASING.md:13]: I'll stay with third person, but additional explanation is always welcome.

File: .github/workflows/release.yaml (hunk 7 above)
Comment (kubawach) [reply .github/workflows/release.yaml:60]: Done.

**Comment by mateuszrzeszutek** [comment]
Implemented in https://github.com/signalfx/splunk-otel-java-lambda/pull/44

**Pull request closed without merging**